T8010: VPP plugin and session hooks interface

AndriiFullroot opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
Presented a hooks interface that allows setting up PPPoE session creation in the VPP dataplane, control IP route routine, and limiter setup.
Also, added flags that allow creating a "non-dev-ppp" (/dev/ppp) interface and creating DHCPv6/ND without an interface socket (if there is no PPP interface, there is no way to create a socket).
The VPP code has been moved to a separate plugin.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T8010

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->

Compile accel-ppp-ng with enabled hooks and VPP:
```
mkdir build && cd build
cmake -DHAVE_VPP=1 -DHAVE_SESSION_HOOKS=1 ../
make
```

Add the vpp plugin and enable the vpp control plane for the interface in the accel-ppp-ng config file:
```
[modules]
pppoe
vpp
...

[pppoe]
interface=eth0,vpp-cp=true
...
```

Make sure that there is an interface passed to VPP, startup.conf:
```
...
dpdk {
    dev 0000:00:00.0
    dev 0000:06:00.0 {
        name eth0
        }
    uio-bind-force
}
...
pppoe {
    enable-pass-nd-and-dhcpv6
}
...
```

Enable PPPoE control plane:
```
vpp# lcp create eth0 host-if eth0
vpp# create pppoe map dp eth0 cp tap4096
```

Connect the client and check that there is a session in the VPP dataplane:
```
vpp# show pppoe session 
Number of PPPoE sessions: 1
[61] sw-if-index 64 client-ip4 0.0.0.0 client-ip6 0.0.0.0/0 session-id 21632 encap-if-index 1 decap-fib-index 0
    local-mac 52:54:00:2f:c4:b4  client-mac 52:54:00:7b:f1:d5
```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

AndriiFullroot committed:
T8010: Added hooks interface

The hook interface is mostly presented for the VPP plugin.
The hook allows us to create a PPP interface, configure iputils, and set
up a limiter.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: DHCPv6 and ND with non-socket packets

Refactoring DHCPv6 and ND to use the session's hooks flags to set up
non-socket packet usage.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: PPP non-dev-ppp option

Refactored PPP to use the session's hooks flags to set up and use
non-dev-ppp(/dev/ppp) code and omit PPP device creation.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: accel_iputils refactoring

Refactored to use the session's hooks for iputils calls.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: ap_session refactoring

Removed vpp-related context from ap_session/ppp.
Refactored to use hooks.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: PPPoE plugin refactoring

Added vpp hook initialization for vpp sessions.
Removed handler check/callback for vpp socket.
Refactored to use hook calls.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: limiter plugin refactoring

Refactored to use hook calls.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: added vpp plugin

Moved all VPP code to a separate plugin.
Implemented VPP plugin hooks.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

ritika0313 commented on accel-pppd/ppp/ppp.c:
```diff
@@ -141,9 +145,10 @@ int __export establish_ppp(struct ppp_t *ppp)
 
 exit_close_chan:
-	if (!ppp->is_vpppoe) {
+#ifdef HAVE_SESSION_HOOKS
+	if (!(ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp))
+#endif /* HAVE_SESSION_HOOKS */
 		close(ppp->chan_fd);
```
This section of code deviates from the previous behavior. 
If HAVE_SESSION_HOOKS is not defined then chan_fd will be closed unconditionally. Previously, the chan_fd was not closed in case of VPP non-existence, probably to avoid double close() because ppp->fd was being closed by the callers of establish_ppp() in case error is returned.
If the intent is not to change the behavior, we might need to move the close statement inside the #ifdef block.

AndriiFullroot replied:
Here we "need" to close the `chan_fd` - but in case of VPP/non-dev-ppp, there is no channel interface and essentially `chan_fd = fd = socket(..PPPoE)`(opened by pppoe plugin) and we need to leave it open. So if there is no "HAVE_SESSION_HOOKS", we need to close chan_fd unconditionally, with hooks - we need to check if it is NOT VPP/non-dev-ppp and then close it.

ritika0313 commented on accel-pppd/ppp/ppp.c:
```diff
@@ -235,7 +255,9 @@ int __export connect_ppp_channel(struct ppp_t *ppp)
 
 exit_close_unit:
-	if (!ppp->is_vpppoe)
+#ifdef HAVE_SESSION_HOOKS
+	if (!(ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp))
+#endif /* HAVE_SESSION_HOOKS */
 		close(ppp->unit_fd);
```
This deviates from the previous behavior too. Please check the intended behavior here and if we should close unit_fd here or not in case HAVE_SESSION_HOOKS is not defined.

AndriiFullroot replied:
We need to leave unit_fd in case of VPP/non-dev-ppp. Because there is no unit /dev/ppp, and effectively, unit_fd is a socket the needs to be open.

ritika0313 commented on accel-pppd/ap_session_hooks.c:
```diff
@@ -0,0 +1,45 @@
+}
+
+__export
+void ap_session_hooks_unregister(struct ap_session_hooks_t *h)
```
Might want to check and return error in case of any NULL pointers passed to this function

AndriiFullroot replied:
Ok

ritika0313 commented on accel-pppd/session_hooks/vyos_vpp/vpputils.c:
```diff
@@ -69,16 +69,18 @@ void vpp_unlock() {
 }
 
-void __export vpp_get()
+int vpp_get()
```
Is there any particular reason to switch to int return type? Currently there doesn't seem a need to return a value.

AndriiFullroot replied:
Well, outside of VPP, the `get()` functions are supposed to initialize hook contexts(if any). I believe it's a good idea to make an int return type.

If there is some issue with VPP(crashed, etc.). There would be a netlink event that would remove all VPP-related PPPoE services.

ritika0313 commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -238,23 +233,18 @@ static void ppp_started(struct ap_session *ses)
-	}
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->pppoe_terminate)
+		ses->hooks->pppoe_terminate(ses, hard);
```
{ses->hooks->pppoe_terminate = vpphook_pppoe_terminate()} -> vpppolicer_remove_limiter() and vpppoe_sync_del_pppoe_interface()
Return value from vpphook_pppoe_terminate() shall be checked here to log any error if occurred while VPP termination.

AndriiFullroot replied:
Yep, that's a good idea.

ritika0313 commented on accel-pppd/session_hooks/vpp/vpphooks.c:
```diff
@@ -0,0 +1,197 @@
+	vpppolicer_remove_limiter(ses);
+	vpppoe_sync_del_pppoe_interface(mac, sid);
+
+	return 0;
```
vpppolicer_remove_limiter() and vpppoe_sync_del_pppoe_interface() can return an error value, you might want to catch them here and return them.

AndriiFullroot replied:
Well, considering that it is a part of pppoe teardown routine, there is not much we can do. But I'll add logs.

ritika0313 commented on accel-pppd/ap_session_hooks.c:
```diff
@@ -0,0 +1,45 @@
+}
+
+__export
+void ap_session_hooks_unregister(struct ap_session_hooks_t *h)
```
Please check if we need to call this function from a place.. may be during PPPoE shutdown for a proper cleanup as currently I don't see a calling to this function

AndriiFullroot replied:
Nope, potentially we need to call it on the VPP plugin "destruction". But in the case of accel-pppd, there is no proper destruction routine for plugins/threads. Potentially, this function would never be called. :(

ritika0313 commented on accel-pppd/ipv6/dhcpv6.c:
```diff
@@ -27,9 +27,7 @@
-#ifdef HAVE_VPP
-# include "vppipv6layer.h"
-#endif /* HAVE_VPP */
+#include "ppp_ipv6layer.h"
```
Do we need to access the contents of this header file for other than VPP functionality ? If not, we can keep the #ifdef HAVE_SESSION_HOOKS enclosures.

AndriiFullroot replied:
Oh yeah, it is better to enclose it.

ritika0313 reviewed (changes requested):
The changes look good covering the necessary changes to expose a session hook layer for VPP by AP sessions. 
Few minor things might need to be checked like the FD closure and a few more (indicated in individual comments).

AndriiFullroot commented:
I've updated the PR and refactored some requests.
Most requests from Sonar I've omitted - IMO, some requested changes to already existing original code, better not to change, at least for now.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -1781,6 +1754,16 @@ void __export pppoe_get_stat(unsigned int **starting, unsigned int **active)
 
+void __export pppoe_get_session_mac_and_sid(struct ap_session *ses, uint8_t **mac, uint16_t *sid)
+{
+	/* AND_TODO: add check if ses is pppoe session? */
```
That would be awesome :3, unfortunately, there is no "protocol" field in the session context.

ritika0313 commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -238,23 +233,19 @@ static void ppp_started(struct ap_session *ses)
-	}
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->pppoe_terminate &&
+		ses->hooks->pppoe_terminate(ses, hard))
```
I meant to capture the return value of 'ses->hooks->pppoe_terminate()' and log an error if its an error value (non-zero)

AndriiFullroot replied:
I'm not sure what the point of logging error values is. As a part of pppoe plugin, this code is explicitly "not" aware of VPP, so the error value doesn't make any sense IMO.

ritika0313 replied:
The idea behind is to log some error in case any error occurs. It is more specific to logging 'some' error than the error value. If you feel that there is no harm to have a silent error returned from VPP in this scenario, please feel free to disregard the comment.

AndriiFullroot replied:
Well, there is a log, but without the error code value:
```
log_ppp_warn("pppoe: issue to terminate PPPoE session with hook\n");
```

ritika0313 replied:
Sorry for the confusion, I missed that you have enclosed the function call inside if(). Thank you for putting the log.

ritika0313 commented on accel-pppd/session_hooks/vyos_vpp/vpphooks.c:
```diff
@@ -0,0 +1,203 @@
+	return ret;
+}
+
+int vpphook_ipaddr_add(struct ap_session *ses, int ifindex, in_addr_t addr, int mask)
```
Please see if we really need to align the argument list with the functions of iputils.c, even if we are not using them in the vpphook functions, or we can get rid of the unused parameters in all these functions (unless we expect a possibility for them to be used in future).

AndriiFullroot replied:
Yes, we don't need g.e. interface index for VPP routes(there is no PPP device, so there is no ifindex), but I believe it's better to present all available arguments for route routines for hooks.

AndriiFullroot committed:
T8010: ap_session/ppp cleanup fixup

Added proper ppp destablish routine for non-dev-ppp.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -238,23 +233,19 @@ static void ppp_started(struct ap_session *ses)
-	}
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->pppoe_terminate &&
+		ses->hooks->pppoe_terminate(ses, hard))
```
Here we are trying to call VPP and destroy VPP's related pppoe_session through the hook.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -479,12 +438,29 @@ static struct pppoe_conn_t *allocate_channel(struct pppoe_serv_t *serv, const ui
-		conn->ppp.ses.non_dev_ppp_fixup = pppoe_create_vpp_session_interface;
+		conn->ppp.ses.hooks = serv->ses_hooks;
+		if (conn->ppp.ses.hooks->session_hook_init &&
+			conn->ppp.ses.hooks->session_hook_init(&conn->ppp.ses)) {
```
Here we are trying to init the hook for a session, mainly allocating memory for the VPP routes list and some memory to store sw_ifindex, etc. The session itself is created in the session.c at the "final" stage after all layers are inited.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -1584,15 +1560,31 @@ static void __pppoe_server_start(const char *ifname, const char *opt, void *cli,
-		vpp_register_handler(&serv->vpp_handler);
-		vpp_get();
+		struct ap_session_hooks_t *ses_hooks = NULL;
+		ses_hooks = ap_session_hooks_find("vpp");
```
If the PPPoE service is started with `vpp-cp=1` argument, then here we are trying to find the hook that implements "vpp" and store it in the serv context so that future connections could initialize it.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
```diff
@@ -1781,6 +1769,16 @@ void __export pppoe_get_stat(unsigned int **starting, unsigned int **active)
 	*active = &stat_active;
 }
 
+void __export pppoe_get_session_mac_and_sid(struct ap_session *ses, uint8_t **mac, uint16_t *sid)
```
This export function is required for the VPP plugin. We need to get Mac and Sid for VPP calls. Mainly, removing the VPP PPPoE session.

AndriiFullroot commented on accel-pppd/include/ap_session.h:
```diff
@@ -124,16 +125,13 @@ struct ap_session
-	int (*vpp_nd_recv)(struct ap_session *, const void *, size_t, struct in6_addr *);
-	int (*vpp_dhcpv6_recv)(struct ap_session *, const void *, size_t, struct in6_addr *, unsigned short);
+	int (*ppp_ipv6_nd_recv)(struct ap_session *, const void *, size_t, struct in6_addr *);
+	int (*ppp_ipv6_dhcpv6_recv)(struct ap_session *, const void *, size_t, struct in6_addr *, unsigned short);
```
Those function pointers are required for the "non-socket" ND/DHCPv6 routine. ND and DHCPv6 service sets up those callbacks to receive ICMPv6 and UDPv6 packets.

AndriiFullroot commented on accel-pppd/ipv6/dhcpv6.c:
Changes in `nd.c` and `dhcpv6.c` are similar. If it is a "non-socket" service, then we need to set up a callback to receive the packets from pppoe socket(through the PPP channel handler in `ppp.c`) and send them through ppp unit(`ppp_ipv6layer.c`).

AndriiFullroot commented on accel-pppd/ipv6/dhcpv6.c:
```diff
@@ -512,11 +515,11 @@ static void dhcpv6_send_reply(struct dhcpv6_packet *req, struct dhcpv6_pd *pd, i
-	if (ses->non_dev_ppp_fixup != NULL) {
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->is_non_socket_dhcpv6_nd) {
 		ipv6layer_unit_dhcpv6_send(ses, reply->hdr, reply->endptr - (void *)reply->hdr, (struct sockaddr *)&req->addr, sizeof(req->addr));
```
Send through ppp unit socket.

AndriiFullroot commented on accel-pppd/ipv6/nd.c:
Changes in `nd.c` and `dhcpv6.c` are similar. If it is a "non-socket" service, then we need to set up a callback to receive the packets from pppoe socket(through the PPP channel handler in `ppp.c`) and send them through ppp unit(`ppp_ipv6layer.c`).

AndriiFullroot commented on accel-pppd/ipv6/ppp_ipv6layer.c:
Set up an IPv6 handler for the unit socket, so we may receive the ICMPv6/UDPv6 packets required for ND/DHCPv6 services.

AndriiFullroot commented on accel-pppd/ipv6/ppp_ipv6layer.c:
Send an ICMPv6 packet, required for "non-socket" ND service.

AndriiFullroot commented on accel-pppd/ipv6/ppp_ipv6layer.c:
Send a DHCPv6 packet through the unit's PPP socket, required for "non-socket" DHCPv6 service.

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
The main idea is to implement the creation of ppp context without the creation of the /dev/ppp interface(aka "non-dev-ppp"). In this case, it means that pppoe socket(opened by `pppoe.c` service) would be set up for chan and unit, aka chan_fd = unit_fd = ppp->fd. Also, the unit and channel read handlers are combined in one.

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
```diff
@@ -118,10 +120,12 @@ int __export establish_ppp(struct ppp_t *ppp)
 	/* Use combined read for non-dev-ppp*/
-	if (ppp->is_vpppoe)
+	if (ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp)
 		ppp->chan_hnd.read = ppp_chan_and_unit_read;
```
Set up a "combined" handler for chan and unit. To enable unit reads, the flag `ppp->is_unit_read_enabled` should be set up.

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
```diff
@@ -83,11 +82,14 @@ int __export establish_ppp(struct ppp_t *ppp)
+#ifdef HAVE_SESSION_HOOKS
+	if (ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp) {
 		/* use fd as chan_fd mostly for sending chan packets */
 		ppp->chan_fd = ppp->fd;
```
`chan_fd = ppp->fd` for "non-dev-ppp".

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
Combined chan/unit read for "non-dev-ppp".

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
```diff
@@ -172,9 +185,12 @@ int __export connect_ppp_channel(struct ppp_t *ppp)
-	if (ppp->is_vpppoe) {
+#ifdef HAVE_SESSION_HOOKS
+	if (ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp) {
 		ppp->unit_fd = ppp->fd;
```
`unit_fd = ppp->fd` for "non-dev-ppp".

AndriiFullroot commented on accel-pppd/ppp/ppp.c:
```diff
@@ -198,10 +214,13 @@ int __export connect_ppp_channel(struct ppp_t *ppp)
+#ifdef HAVE_SESSION_HOOKS
+	if (ppp->ses.hooks && ppp->ses.hooks->is_non_dev_ppp) {
 		ppp->ses.ifindex = -1;
 		ppp->is_unit_read_enabled = 1;
```
Set up unit read handlers to be active in "combined" chan/unit read for "non-dev-ppp".

AndriiFullroot commented on accel-pppd/ppp/ppp.h:
Possible handler for "non-socket" ND/DHCPv6 services. Handler allows reception of IPv6 ICMP/UDP packets.

AndriiFullroot commented on accel-pppd/shaper/limiter.c:
```diff
@@ -542,15 +540,15 @@ static int remove_htb_ifb(struct rtnl_handle *rth, int ifindex, int priority)
 		up_speed = up_speed * 1000 / 8;
 		up_burst = up_burst ? up_burst : conf_up_burst_factor * up_speed;
-		return vpppolicer_install_limiter(ses, down_speed, down_burst, up_speed, up_burst);
+		return ses->hooks->install_limiter(ses, down_speed, down_burst, up_speed, up_burst);
```
Hook - use VPP limiter instead of the kernel's.

AndriiFullroot commented on accel-pppd/shaper/limiter.c:
```diff
@@ -608,11 +606,10 @@ int install_limiter(struct ap_session *ses, int down_speed, int down_burst, int
-#endif
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->remove_limiter)
+		return ses->hooks->remove_limiter(ses);
```
Hook - remove VPP limiter instead of the kernel's.

AndriiFullroot commented on accel-pppd/accel_iputils.c:
This file implements all iputil-related calls (ip/route routines) and checks for a hook to call VPP's or the kernel/netlink routine.

AndriiFullroot commented on accel-pppd/session.c:
```diff
@@ -159,12 +164,14 @@ void __export ap_session_activate(struct ap_session *ses)
-		if (ses->non_dev_ppp_fixup(ses)) {
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks != NULL && ses->hooks->pppoe_create_session_interface) {
+		if (ses->hooks->pppoe_create_session_interface(ses)) {
```
Here we are trying to create "pppoe_session" in VPP on the "final" step of PPPoE initialization. So, the PPPoE session would be created after all layers (authentication, IPCP, etc) are done.

AndriiFullroot commented on accel-pppd/session.c:
```diff
@@ -289,6 +296,11 @@ void __export ap_session_finished(struct ap_session *ses)
+
+#ifdef HAVE_SESSION_HOOKS
+	if (ses->hooks && ses->hooks->session_hook_deinit)
+		ses->hooks->session_hook_deinit(ses);
```
Free, hooks private data, aka VPP's route list, and some memory to store(g.e, sw_ifindex).

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
Callback for "new link event". So we may restart the service on the interface.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
Callback for "new link event". If the interface is gone(mainly means that VPP has crashed), then we need to "free" all sessions related to the service.

AndriiFullroot commented on accel-pppd/ctrl/pppoe/pppoe.c:
Register "new link event" handler.

AndriiFullroot committed:
T8010: Rename vpp plugin to vyos_vpp plugin

Renamed, considering that the current accel-ppp VPP
integration works only with VyOS's VPP version.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>

AndriiFullroot committed:
T8010: Refactoring VPP hooks private data

Using the session's private data list instead of a dedicated field.

Signed-off-by: Andrii Melnychenko <a.melnychenko@vyos.io>
